### PR TITLE
feat: end loop

### PR DESCRIPTION
**Pull request by Kylescottw**

### CODE REVIEW

**Comment by henriquehbr** [comment]
Thanks for your valuable contribution @Kylescottw!

Just a little consideration before we can get things going, could you please include a little description of what actually this PR aims to bring as a addition? (as it's labelled as a feature, in this case), if possible, some code or visual demonstration through GIF's or images is always really useful for having a better way to show of your idea in a more practical and comprehensive way

Once again, thanks for your time and effort on this :)

**Comment by Kylescottw** [comment]
Sure thing!

**Comment by Kylescottw** [comment]
The feature allows the loop to run through all elements and stops on the last item. 
[](url)

https://user-images.githubusercontent.com/14112830/110268735-71197500-7f7f-11eb-9aeb-600e2a8c460f.mov

**Comment by henriquehbr** [comment]
This is looking amazing, great job!

I'll test it locally just to make sure everything is working fine (and make proper adaptions if necessary), otherwise, if everything is running smoothly, expect to see it shipped soon

**Comment by Kylescottw** [comment]
Sounds great!
Thank you

File: lib/esm/Typewriter.mjs [hunk 1]
```diff
@@ -1 +1 @@
...
...,l&1&&g(e,"cursor",c[0]),l&4&&g(e,"delay",c[2].delay>0)},i(c){o||(b(i,c),o=!0)},o(c){C(i,c),o=!1},d(c){c&&k(e),i&&i.d(c),r=!1,s()}}}function Et(t){let e=t[1],n,o,r=ft(t);return{c(){r.c(),n=K()},m(s,u){r.m(s,u),D(s,n,u),o=!0},p(s,[u]){u&2&&O(e,e=s[1])?(et(),C(r,1,1,h),nt(),r=ft(s),r.c(),b(r),r.m(n.parentNode,n)):r.p(s,u)},i(s){o||(b(r),o=!0)},o(s){C(r),o=!1},d(s){s&&k(n),r.d(s)}}}function St(t,e,n){let{$$slots:o={},$$scope:r}=e,{interval:s=30}=e,{cascade:u=!1}=e,{loop:i=!1}=e,{loopRandom:c=!1}=e,{loopOnce:l=!1}=e,{cursor:_=!0}=e,{delay:f=0}=e,p=!1,m={},w=N();A(()=>p&&n(1,m={})),j(()=>p=!0),t.$$set=a=>{"interval"in a&&n(3,s=a.interval),"cascade"in a&&n(4,u=a.cascade),"loop"in a&&n(5,i=a.loop),"loopRandom"in a&&n(6,c=a.loopRandom),"loopOnce"in a&&n(7,l=a.loopOnce),"cursor"in a&&n(0,_=a.cursor),"delay"in a&&n(8,f=a.delay),"$$scope"in a&&n(9,r=a.$$scope)};let H;return t.$$.update=()=>{if(t.$$.dirty&505){t:n(2,H={interval:s,cascade:u,loop:i,loopRandom:c,loopOnce:l,cursor:_,delay:f,dispatch:w})}},[_,m,H,s,u,i,c,l,f,r,o]}var dt=class extends z{constructor(e){super();rt(this,e,St,Et,O,{interval:3,cascade:4,loop:5,loopRandom:6,loopOnce:7,cursor:0,delay:8})}},$t=dt;at();export{$t as default};
```
Comment (henriquehbr) [inline lib/esm/Typewriter.mjs:1]: As the support for both ESM and UMD bundles has been deprecated on [3.0.0-alpha.0](https://github.com/henriquehbr/svelte-typewriter/blob/main/CHANGELOG.md#300-alpha0---2021-11-05), i think we can remove this

File: lib/esm/modes/loopTypewriter.mjs [hunk 2]
```diff
@@ -1 +1 @@
...
+import{a as f,b as p,c as m}from"../chunk.KEGSSE5H.mjs";import"../chunk.E4LFTDIS.mjs";var a=async(r,e)=>{e.dispatch("done"),await p(typeof e.loop=="number"?e.loop:1500);let t=r.innerHTML.replaceAll("&amp;","&");for(let n=t.length-1;n>=0;n--)t[n]===">"&&(n=t.lastIndexOf("<",n)),r.innerHTML=t.slice(0,n),await p(e.interval)},i=[],s=r=>{for(;;){let e=f(0,r.length),t=typeof i=="number"&&e!==i;if(Array.isArray(i)&&!i.includes(e)||t)return t&&(i=[]),i.push(e),r[e];i.length===r.length&&(i=i.pop())}};async function c(r,e){for(let t=0;t<r.length;t++)await e(r[t],t,r)}var l=async({currentNode:r,text:e},t)=>{if(await m({currentNode:r,text:e},t),!t.stopLoop){let n=e.replaceAll("&","&amp;");r.innerHTML===n&&await a(r,t)}},y=async(r,e)=>{if(e.loopOnce)c(r,async(t,n)=>{let o=e.loopOnce?r.length-1===n:!1;await l(t,{...e,stopLoop:o})});else for(;;)if(e.loop)for(let t of r)await l(t,e);else if(e.loopRandom){let t=s(r);await l(t,e)}};export{y as mode};
```
Comment (henriquehbr) [inline lib/esm/modes/loopTypewriter.mjs:1]: Same as https://github.com/henriquehbr/svelte-typewriter/pull/46#discussion_r744148566

File: src/helpers/asyncForEach.ts [hunk 3]
```diff
@@ -0,0 +1,7 @@
+async function asyncForEach(array: any, callback: Function) {
```
Comment (henriquehbr) [inline src/helpers/asyncForEach.ts:1]: This helper should be converted to `.js`, as the project migrated from TS to JS on [3.0.0-alpha.0](https://github.com/henriquehbr/svelte-typewriter/pull/46#discussion_r744148566)

File: example/src/pages/Loop.svelte [hunk 4]
```diff
@@ -54,7 +65,7 @@
   <slot />
 </TypewriterControls>
 
-<hr>
+<hr>
```
Comment (henriquehbr) [inline example/src/pages/Loop.svelte:68]: This trailling whitespace should be removed

File: example/src/pages/Loop.svelte [hunk 5]
```diff
@@ -14,6 +14,17 @@
...
+
+<pre>
+  &lt;Typewriter loop loopOnce /&gt;
+</pre>
+
+<TypewriterControls loop loopOnce>
+  <slot />
+</TypewriterControls>
+
+
```
Comment (henriquehbr) [inline example/src/pages/Loop.svelte:27]: This second line break should be removed

File: example/src/pages/Loop.svelte [hunk 6]
```diff
@@ -5,7 +5,7 @@
 <h1>Loop mode</h1>
 
 <hr>
-
+
```
Comment (henriquehbr) [inline example/src/pages/Loop.svelte:8]: Same as https://github.com/henriquehbr/svelte-typewriter/pull/46#discussion_r744149184

File: lib/cjs/Typewriter.js [hunk 7]
```diff
@@ -1 +1 @@
...
...pdate.call(null,c[2]),l&1&&v(t,"cursor",c[0]),l&4&&v(t,"delay",c[2].delay>0)},i(c){r||(x(s,c),r=!0)},o(c){R(s,c),r=!1},d(c){c&&O(t),s&&s.d(c),o=!1,i()}}}function st(e){let t=e[1],n,r,o=Oe(e);return{c(){o.c(),n=le()},m(i,u){o.m(i,u),P(i,n,u),r=!0},p(i,[u]){u&2&&N(t,t=i[1])?(de(),R(o,1,1,h),pe(),o=Oe(i),o.c(),x(o),o.m(n.parentNode,n)):o.p(i,u)},i(i){r||(x(o),r=!0)},o(i){R(o),r=!1},d(i){i&&O(n),o.d(i)}}}function lt(e,t,n){let{$$slots:r={},$$scope:o}=t,{interval:i=30}=t,{cascade:u=!1}=t,{loop:s=!1}=t,{loopRandom:c=!1}=t,{loopOnce:l=!1}=t,{cursor:m=!0}=t,{delay:f=0}=t,_=!1,y={},S=U();q(()=>_&&n(1,y={})),z(()=>_=!0),e.$$set=a=>{"interval"in a&&n(3,i=a.interval),"cascade"in a&&n(4,u=a.cascade),"loop"in a&&n(5,s=a.loop),"loopRandom"in a&&n(6,c=a.loopRandom),"loopOnce"in a&&n(7,l=a.loopOnce),"cursor"in a&&n(0,m=a.cursor),"delay"in a&&n(8,f=a.delay),"$$scope"in a&&n(9,o=a.$$scope)};let Q;return e.$$.update=()=>{if(e.$$.dirty&505){e:n(2,Q={interval:i,cascade:u,loop:s,loopRandom:c,loopOnce:l,cursor:m,delay:f,dispatch:S})}},[m,y,Q,i,u,s,c,l,f,o,r]}var Ae=class extends J{constructor(t){super();he(this,t,lt,st,N,{interval:3,cascade:4,loop:5,loopRandom:6,loopOnce:7,cursor:0,delay:8})}},ct=Ae;$e();
```
Comment (henriquehbr) [inline lib/cjs/Typewriter.js:1]: Same as https://github.com/henriquehbr/svelte-typewriter/pull/46#discussion_r744148566

File: example/src/pages/Loop.svelte [hunk 8]
```diff
@@ -14,6 +14,17 @@
   <slot />
 </TypewriterControls>
 
+<hr> 
+
+<pre>
+  &lt;Typewriter loop loopOnce /&gt;
+</pre>
+
+<TypewriterControls loop loopOnce>
```
Comment (henriquehbr) [inline example/src/pages/Loop.svelte:23]: Shouldn't only `loopOnce` be passed here?

File: src/modes/loopTypewriter.ts [hunk 9]
```diff
@@ -1,22 +1,32 @@
...
+		while (true) {
+			if (options.loop) {
+				for (const element of elements) await loopTypewriter(element, options)
+			} else if (options.loopRandom) {
+				const element = getRandomElement(elements)
+				await loopTypewriter(element, options)
+			}
 		}
 	}
 }
```
Comment (henriquehbr) [inline src/modes/loopTypewriter.ts:32]: Same as https://github.com/henriquehbr/svelte-typewriter/pull/46#discussion_r744148787

File: src/helpers/asyncForEach.ts [hunk 10]
```diff
@@ -0,0 +1,7 @@
...
+	for (let index = 0; index < array.length; index++) {
+		await callback(array[index], index, array)
+	}
+}
+
+export { asyncForEach }
```
Comment (henriquehbr) [inline src/helpers/asyncForEach.ts:7]: This helper should be converted to `.js`, as the project migrated from TS to JS on [3.0.0-alpha.0](https://github.com/henriquehbr/svelte-typewriter/pull/46#discussion_r744148566)

**Comment by henriquehbr** [comment]
As this feature has been introduced on version [3.0.0-alpha.18](https://github.com/henriquehbr/svelte-typewriter/blob/main/CHANGELOG.md#300-alpha18---2022-07-02), i'll be closing this PR, many thanks for your idea and contribution!

**Pull request closed without merging**